docker/dockerfile_build.bzl

```diff
@@ -0,0 +1,113 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Repository rule for building a Docker image from a Dockerfile and saving it
+   into a tar file.
+
+This rule uses the docker tool installed in the system to build the image and
+then save it as a tar file. The saved tar file is available for other rules to
+use (e.g. container_image, container_load, container_test)
+
+The produced tar file is available at @<repo_name>//image:dockerfile_image.tar
+"""
+
+def docker(repository_ctx):
+    """Resolves the docker path.
+
+    Args:
+      repository_ctx: The repository context
+
+    Returns:
+      Tha path to the docker tool
+    """
+
+    if repository_ctx.attr.docker_path:
+        return repository_ctx.attr.docker_path
+    if repository_ctx.which("docker"):
+        return repository_ctx.which("docker")
+
+    fail("Path to the docker tool was not provided and it could not be resolved automatically.")
+
+def _impl(repository_ctx):
+    """Core implementation of dockerfile_image."""
+
+    docker_path = docker(repository_ctx)
+
+    dockerfile_src = repository_ctx.path(repository_ctx.attr.dockerfile)
+    dockerfile_dest = repository_ctx.path("Dockerfile")
+
+    # Copy the provided Dockerfile into the root of this workspace since using
+    # the -f flag with the docker build command is problematic.
+    cp_result = repository_ctx.execute([
+        "cp",
+        dockerfile_src,
+        dockerfile_dest,
+    ])
+    if cp_result.return_code:
+        fail("Failed to copy Dockerfile from {} to {}: {}".format(
+            dockerfile_src,
+            dockerfile_dest,
+            cp_result.stderr,
+        ))
+
+    img_name = repository_ctx.name + ":dockerfile_image"
+    build_args = [
+        docker_path,
```

Comment:
We need to add arguments to the docker call so that it DOES not use the docker cache (i.e., it always builds the full image using the DOCKERFILE from scratch) (Pointer I found: https://stackoverflow.com/questions/35594987/how-to-force-docker-for-a-clean-build-of-an-image)

Reply:
Added the `--no-cache` flag.

Reply:
Sorry for bumping an old thread, but I want to understand the reasoning behind not using the docker cache. I was under impression that docker builds are hermetic, since they are cached based on the file content hashes. Is that not true?

Reply:
One big reason is that a Dockerfile allows specifying arbitrary commands inside the RUN instruction. This implies that the command under the RUN instruction may produce different results when running multiple times, but using the docker cache will prevent a new image to be built. For example, you could be installing some apt package foo with
`apt-get install foo`
and that package gets updated in 5 mins. You rebuild your image, but won't get the updated package since your layers will be cached.

Reply:
Thanks, that makes sense. I wouldn't mind having some way to control this behavior though, it's a pain in the neck when developing stuff locally.

Reply:
@f355, thanks for your comments. The big argument against using the Docker cache is that Bazel has its own caching mechanism. Getting the two to play along nicely is not possible afaik. The most likely result is very frustrating situations where even running `bazel clean --expunge_async` yields non-hermetic / non-deterministic behavior.
If by 'pain in the neck' you mean that stuff runs a bit slowly, I'd take that every day vs having users complain about hermeticity problems (i.e., why does the test run in my machine but fails in yours?). We've seen this pain with docker images and wish it upon noone. If you feel building with a Dockerfile is too slow, our advice is to start migrating it to rules_docker to get better performance.